### PR TITLE
Add username replacement API

### PR DESCRIPTION
**Pull request by tuchfarber**
New API replaces all copies of username in this service with a new
username. Requires user be in a username_replacement_admin group. Should
only be ran as a larger job to update usernames across all services,
otherwise the system will be left in a broken state for those users.

Credentials Pull Request
---

Make sure that the following steps are done before rebasing/merging

  - [ ] Request a review if desired
  - [ ] Squash/Fixup your branch to one commit
  - Config/Dependency Changes (e.g. config files, scripts that are used for provisioning etc.)
    - [ ] Make sure you have updated [edx/configuration](https://github.com/edx/configuration) and [edx/devstack](https://github.com/edx/devstack) if necessary
    - [ ] Make sure the change builds successfully in a sandbox
  - UI Changes 
    - [ ] Consider other browsers (e.g. Firefox)
    - [ ] Check mobile view
    - [ ] Consider accessibility (e.g. Run aXe on the page)

### CODE REVIEW

File: credentials/apps/api/v2/views.py [hunk 1]
```diff
@@ -145,3 +148,135 @@ def partial_update(self, request, *args, **kwargs):  # pylint: disable=useless-s
...
+                return False
+        return True
+
+    def _replace_username_for_all_models(self, current_username, new_username, replacement_locations):
+        """
+        Replaces current_username with new_username for all (model, column) pairs in replacement locations.
+        Returns if it was successful or not. Will return successful even if no matching
+
+        TODO: Determine if logs of username are a PII issue.
+        """
```
Comment (BessieSteinberg) [inline credentials/apps/api/v2/views.py]: Remove TODO

File: credentials/apps/api/v2/tests/test_views.py [hunk 2]
```diff
@@ -517,3 +519,82 @@ def test_throttle_configuration(self, scope):
...
+        existing_users = [{user.username: user.username + '_new'} for user in random_users]
+        non_existing_users = [{username: username + '_new'} for username in fake_usernames]
+        data = {
+            "username_mappings": existing_users + non_existing_users
+        }
+        expected_response = {
+            'failed_replacements': [],
+            'successful_replacements': existing_users + non_existing_users
+        }
+        response = self.call_api(self.service_user, data)
```
Comment (BessieSteinberg) [inline credentials/apps/api/v2/tests/test_views.py]: Same comment as on discovery: Is the expected behavior that there will be successful replacements for non-existing users? Are new user records created for users if the old username doesn't exist? It's fine if that is the case just add that to the docstring for the api and maybe add a comment about that in this test

Comment (tuchfarber) [reply credentials/apps/api/v2/tests/test_views.py]: I'll definitely add some more comments around this in the services as this was something I realized later and changed. If a user never views a credential, they aren't going to exist on the credentials service at all. So we should treat the lack of a user a success as we don't need to do any work.

File: credentials/apps/api/v2/views.py (hunk 1 above)
Comment (tuchfarber) [reply credentials/apps/api/v2/views.py]: Removed

File: credentials/apps/api/v2/views.py [hunk 3]
```diff
@@ -145,3 +148,135 @@ def partial_update(self, request, *args, **kwargs):  # pylint: disable=useless-s
...
         """ Update a grade. """
         return super(GradeViewSet, self).update(request, *args, **kwargs)
+
+
+class UsernameReplacementView(APIView):
+    """
+    WARNING: This API is only meant to be used as part of a larger job that
+    updates usernames across all services. DO NOT run this alone or users will
+    not match across the system and things will be broken.
+
```
[the remark this answers is not in the document]
Comment (tuchfarber) [reply credentials/apps/api/v2/views.py]: Added note.

File: credentials/apps/api/v2/views.py [hunk 4]
```diff
@@ -145,3 +148,136 @@ def partial_update(self, request, *args, **kwargs):  # pylint: disable=useless-s
...
+        {
+            "successful_replacements": [
+                {"old_username_1": "new_username_1"}
+            ],
+            "failed_replacements": [
+                {"old_username_2": "new_username_2"}
+            ]
+        }
+
+        TODO: Determine if we need an audit trail outside of logging and API response.
```
Comment (georgebabey) [inline credentials/apps/api/v2/views.py]: Is this TODO done? Do we need anything further?

File: credentials/apps/api/v2/permissions.py [hunk 5]
```diff
@@ -40,3 +41,12 @@ def has_permission(self, request, view):
             return default or (request.user.username.lower() == filtered_username)
 
         return default
+
+
+class CanReplaceUsername(permissions.BasePermission):
+    """
+    Grants access to the Username Replacement API for anyone in the group,
```
Comment (georgebabey) [inline credentials/apps/api/v2/permissions.py]: remove mention of group

**Pull request merged**